demos/http/http_demo_s3_download_async/http_demo_s3_download_async.c

```diff
@@ -0,0 +1,751 @@
+
+/**
+ * @brief Delay in seconds between each iteration of the demo.
+ */
+#define DEMO_LOOP_DELAY_SECONDS        ( 5U )
+
+/* Check that transport timeout for transport send and receive is defined. */
+#ifndef TRANSPORT_SEND_RECV_TIMEOUT_MS
+    #define TRANSPORT_SEND_RECV_TIMEOUT_MS    ( 1000 )
+#endif
+
+/* Check that size of the user buffer is defined. */
+#ifndef USER_BUFFER_LENGTH
+    #define USER_BUFFER_LENGTH    ( 4096 )
+#endif
+
+/* Check that request queue name is defined. */
+#ifndef REQUEST_QUEUE
+    #define REQUEST_QUEUE    "/demo_request_queue"
+#endif
+
+/* Check that request queue name is defined. */
+#ifndef RESPONSE_QUEUE
+    #define RESPONSE_QUEUE    "/demo_response_queue"
+#endif
+
+/**
+ * @brief Field name of the HTTP Range header to read from server response.
+ */
+#define HTTP_CONTENT_RANGE_HEADER_FIELD           "Content-Range"
+
+/**
+ * @brief Length of the HTTP Range header field.
+ */
+#define HTTP_CONTENT_RANGE_HEADER_FIELD_LENGTH    ( sizeof( HTTP_CONTENT_RANGE_HEADER_FIELD ) - 1 )
+
+/* The location of the host address within string S3_PRESIGNED_URL. */
+static const char * pAddress = NULL;
+
+/* The host address string extracted from S3_PRESIGNED_URL. */
+static char serverHost[ MAX_HOST_ADRESS_LENGTH ];
+
+/* The length of the host address found in string S3_PRESIGNED_URL. */
+static size_t serverHostLength = 0;
+
+/**
+ * @brief Data type for request queue.
+ *
+ * In addition to sending the headers struct, we need to send the buffer it uses.
+ * The pointer to the buffer in the headers struct will be incorrect after
+ * it is received so will need to be fixed.
+ */
+typedef struct RequestItem
+{
+    HTTPRequestHeaders_t requestHeaders;
+    uint8_t headersBuffer[ USER_BUFFER_LENGTH ];
+} RequestItem_t;
+
+/**
+ * @brief Data type for response queue.
+ *
+ * In addition to sending the response struct, we need to send the buffer it uses.
+ * The pointer to the buffer in the response struct will be incorrect after
+ * it is received so will need to be fixed.
+ */
+typedef struct ResponseItem
+{
+    HTTPResponse_t response;
+    uint8_t responseBuffer[ USER_BUFFER_LENGTH ];
+} ResponseItem_t;
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Connect to HTTP server with reconnection retries.
+ *
+ * @param[out] pNetworkContext The output parameter to return the created network context.
+ *
+ * @return EXIT_FAILURE on failure; EXIT_SUCCESS on successful connection.
+ */
+static int32_t connectToServer( NetworkContext_t * pNetworkContext );
+
+/**
+ * @brief Enqueue HTTP requests based on a specified method and path.
+ *
+ * @param[in] pHost The host name of the server.
+ * @param[in] pMethod The HTTP request method.
+ * @param[in] pPath The Request-URI to the objects of interest.
+ * @param[in] requestQueue The queue to write requests to.
+ * @param[out] requestCount The number of HTTP requests queued.
+ *
+ * @return EXIT_FAILURE on failure; EXIT_SUCCESS on success.
```

Comment:
Can we not return EXIT_FAILURE or EXIT_SUCCESS from private functions and instead use bool or HTTPStatus_t here?